File: 2178.py
from collections import deque
n, m = map(int, input().split())
miro = [list(map(int, input().strip())) for _ in range(n)]
dist = [[0 for _ in range(m)] for _ in range(n)]
visit = [[False for _ in range(m)] for _ in range(n)]
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]
#(dx[i], dy[i])는 순서대로 오른쪽, 왼쪽, 아래, 위

#BFS start
que = deque()
que.append((0, 0))
visit[0][0] = True
dist[0][0] = 1

while que :
    ax, ay = que.popleft()
    for i in range(4) :
        new_x = ax + dx[i]
        new_y = ay + dy[i]

        if 0 <= new_x < n and 0 <= new_y < m :
            if visit[new_x][new_y] == False and miro[new_x][new_y] == 1 :
                que.append((new_x, new_y))
                dist[new_x][new_y] = dist[ax][ay] + 1
                visit[new_x][new_y] = True

result = dist[n - 1][m - 1]
print(result)

File: 2178.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#pragma warning (disable: 4996)

using namespace std;

int map[100][100];
bool visit[100][100];
int dist[100][100];
int dx[4] = { 0, 0, 1, -1 };
int dy[4] = { 1, -1, 0, 0 };

int BFS(int N, int M) {
	queue<pair<int, int>> que;
	que.push(pair<int, int>(0, 0));
	visit[0][0] = true;
	dist[0][0] = 1;

	while (!que.empty()) {
		int ax = que.front().first;
		int ay = que.front().second;
		que.pop();

		for (int i = 0; i < 4; i++) {
			int nx = ax + dx[i];
			int ny = ay + dy[i];

			if ((nx >= 0 && nx < N) && (ny >= 0 && ny < M)) {
				if ((visit[nx][ny] == false) && (map[nx][ny] == 1)) {
					que.push(pair<int, int>(nx, ny));
					dist[nx][ny] = dist[ax][ay] + 1;
					visit[nx][ny] = true;
				}
			}
		}
	}
	int result = dist[N - 1][M - 1];

	return result;
}

int main(void) {
	int N, M;
	cin >> N >> M;
	
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			scanf("%1d", &map[i][j]);
		}
	}

	cout << BFS(N, M);
	return 0;
}
